aiida/engine/processes/workchains/restart.py

```diff
@@ -245,23 +249,20 @@ def __init__(self, *args, **kwargs):
         if self._process_class is None or not issubclass(self._process_class, Process):
             raise ValueError('no valid Process class defined for `_process_class` attribute')
 
-    @classproperty
-    def _handlers(cls):  # pylint: disable=no-self-argument
+    def _handlers(self):
         """Return the tuple of all registered handlers for this class and of any parent class.
 
         :return: tuple of handler methods
         """
-        return getattr(super(), '__handlers', tuple()) + cls.__handlers
+        from .utils import process_handler
 
-    @classmethod
-    def register_handler(cls, name, handler):
-        """Register a new handler to this class.
+        handlers = []
 
-        :param name: the name under which to register the handler
-        :param handler: a method with the signature `self, node`.
-        """
-        setattr(cls, name, handler)
-        cls.__handlers = cls.__handlers + (handler,)
+        for method in inspect.getmembers(self, predicate=inspect.ismethod):
+            if hasattr(method[1], 'decorator') and method[1].decorator == process_handler:  # pylint: disable=comparison-with-callable
```

Comment:
what about adding a check that handler is active?

Reply:
This will be implemented in PR for the separate issue #3773